### PR TITLE
File hashes

### LINKED ISSUE #493: c_time vs m_time vs checksum
m_time → ux → important
  Folders don’t have m_times

current state
m_time
Desktop → get time from client
Android → gets time from server

in the dm the c_time replace due hashes

### PR DESCRIPTION
**Pull request by julianseeger**
resolves #493

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
 ## [Unreleased]
 ### CHANGES
 - AbstractNavigation::upload no longer throws QblStorageNameConflictExceptions (they are just resolved)
-- #488 FileNameConflicts are resolved in reverse order. The remote file is renamed now.
+- #488 FileNameConflicts are resolved in reverse order. The remote savedFile is renamed now.
```
Comment (audax) [inline CHANGELOG.md]: I guess this was not intentional.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,10 +1,10 @@
 # Contributing
 
-This file describes the way we work. Reading this file will help you to understand the organization of the Qabel project. If you want to contribute to Qabel -- depending on your experience with open source projects and GitHub -- please scan this file on how to do this.
+This savedFile describes the way we work. Reading this savedFile will help you to understand the organization of the Qabel project. If you want to contribute to Qabel -- depending on your experience with open source projects and GitHub -- please scan this savedFile on how to do this.
 
 ## Who should read this
 
-This file is for contributors, reviewers and integrators alike.
+This savedFile is for contributors, reviewers and integrators alike.
```
Comment (audax) [inline CONTRIBUTING.md]: This savedFile is just a file ;)

File: box/src/main/java/de/qabel/box/storage/AbstractNavigation.kt [hunk 3]
```diff
@@ -175,9 +178,11 @@ abstract class AbstractNavigation(private val prefix: String, protected var dm:
     private fun uploadFile(name: String, file: File, expectedFile: BoxFileState?, listener: ProgressListener?): BoxFile {
         val key = cryptoUtils.generateSymmetricKey()
         val block = UUID.randomUUID().toString()
+
         val oldFile = dm.getFile(name)
+        val shareState = Share.create (oldFile?.meta, oldFile?.metakey)
 
-        val boxFile = BoxFile(prefix, block, name, file.length(), 0L, key.key, oldFile?.meta, oldFile?.metakey)
+        val boxFile = BoxFile(prefix, block, name, file.length(), 0L, key.key, shared = shareState)
```
Comment (audax) [inline box/src/main/java/de/qabel/box/storage/AbstractNavigation.kt]: if shareState isn't used anywhere else, I think it should be inlined.

File: box/src/main/java/de/qabel/box/storage/BoxVolume.kt [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        try {
+            loadDriver()
+        } catch (e: ClassNotFoundException) {
+            throw RuntimeException(e)
+        }
+
+    }
+
+    @Throws(ClassNotFoundException::class)
+    protected fun loadDriver() {
```
Comment (audax) [inline box/src/main/java/de/qabel/box/storage/BoxVolume.kt]: Either the class and this method need to be open, or the sqlite driver loading needs to happen in another class.

File: box/src/main/java/de/qabel/box/storage/BoxVolume.kt [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        val dm = JdbcDirectoryMetadataFactory(tempDir, deviceId).open(tmp, rootRef)
+        return DefaultIndexNavigation(prefix, dm, keyPair, deviceId, readBackend, writeBackend)
+    }
+
+    /**
+     * Calculate the filename of the index metadata file
+
+     * @throws QblStorageException if the hash algorithm could not be found
+     */
+    val rootRef: String
```
Comment (audax) [inline box/src/main/java/de/qabel/box/storage/BoxVolume.kt]: This should be a lazy property

File: box/src/main/java/de/qabel/box/storage/BoxVolume.kt [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            val bb = ByteBuffer.wrap(firstBytes)
+            val uuid = UUID(bb.long, bb.long)
+            return uuid.toString()
+        }
+
+    /**
+     * Create a new index metadata file
+     */
+    @Throws(QblStorageException::class)
+    fun createIndex(bucket: String, prefix: String) {
```
Comment (audax) [inline box/src/main/java/de/qabel/box/storage/BoxVolume.kt]: We don't need this one anymore.

File: box/src/main/java/de/qabel/box/storage/BoxVolume.kt [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        createIndex("https://$bucket.s3.amazonaws.com/$prefix")
+    }
+
+    /**
+     * Create a new index metadata file
+     */
+    @Throws(QblStorageException::class)
+    fun createIndex(root: String) {
+        val rootRef = rootRef
```
Comment (audax) [inline box/src/main/java/de/qabel/box/storage/BoxVolume.kt]: A lazy property could be used directly

File: box/src/test/java/de/qabel/box/storage/testFile.txt [hunk 8]
```diff
@@ -1,3 +1,3 @@
-This a test file to test file encryption for storage module.
+This a test savedFile to test savedFile encryption for storage module.
```
Comment (audax) [inline box/src/test/java/de/qabel/box/storage/testFile.txt]: Well...it is still a sentence :D

File: core.umlcd [hunk 9]
```diff
@@ -5402,7 +5402,7 @@
           <modelElement href="pathmap://UML_LIBRARIES/JavaPrimitiveTypes.library.uml#boolean"/>
         </type>
       </outParameters>
-      <parameters xmi:id="_wvGiMpS-EeSNP8fh6jUNPw" name="file">
+      <parameters xmi:id="_wvGiMpS-EeSNP8fh6jUNPw" name="savedFile">
```
Comment (audax) [inline core.umlcd]: Just delete this file.

File: core/src/resources/config/log4j2.xml [hunk 10]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!-- Simple Logger with Level all, and appender to console and file -->
+<!-- Simple Logger with Level all, and appender to console and savedFile -->
```
Comment (audax) [inline core/src/resources/config/log4j2.xml]: !

**Comment by audax** [comment]
Is this actually this WIP?

**Comment by julianseeger** [comment]
Yip. The file hashes now have a place to be but nobody calculates and sets them.
I'm currently extending the AbstractNavigation to automagically calculate and add hashes when encrypting ...

**Comment by julianseeger** [comment]
Now it's done. Every uploaded file gets a Hash calculated (by default Blake2b) and this hash is inserted into the corresponding DM and FM and shipped with it's BoxFile instance.

**Pull request merged**